### PR TITLE
[Merged by Bors] - feat(RingTheory/DedekindDomain):  `IsDedekindDomainDvr` implies `IsDedekindDomain`

### PR DESCRIPTION
**Pull request by mbkybky**
Prove that `IsDedekindDomainDvr` implies `IsDedekindDomain`, i.e., if an integral domain is Noetherian, and the localization at every nonzero prime is a discrete valuation ring, then it is a Dedekind domain.

Co-authored-by: Yongle Hu @mbkybky
Co-authored-by: Jiedong Jiang @jjdishere

- [x] depends on: #16558

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/DedekindDomain/Dvr.lean [hunk 1]
```diff
@@ -143,3 +141,38 @@ theorem IsDedekindDomain.isDedekindDomainDvr [IsDedekindDomain A] : IsDedekindDo
...
+variable {A}
+
+theorem IsDedekindDomainDvr.ring_dimensionLEOne (h : IsDedekindDomainDvr A) :
+    Ring.DimensionLEOne A where
+  maximalOfPrime := by
+    intro p hp hpp
+    rcases p.exists_le_maximal (Ideal.IsPrime.ne_top hpp) with ⟨q, hq, hpq⟩
+    let f := (IsLocalization.orderIsoOfPrime q.primeCompl (Localization.AtPrime q)).symm
+    let P := f ⟨p, hpp, HasSubset.Subset.disjoint_compl_left hpq⟩
+    let Q := f ⟨q, hq.isPrime, HasSubset.Subset.disjoint_compl_left (fun _ a => a)⟩
```
Comment (Ruben-VandeVelde) [inline Mathlib/RingTheory/DedekindDomain/Dvr.lean:154]: These `HasSubset.Subset` names are only for dot notation; either use them that way or use the theorem they're aliases for, please

**Review by riccardobrasca — COMMENTED** [review]
Is the other implication already in mathlib?

**Comment by mbkybky** [comment]
> Is the other implication already in mathlib?

Yes. It is [`IsDedekindDomain.isDedekindDomainDvr`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/DedekindDomain/Dvr.html#IsDedekindDomain.isDedekindDomainDvr).

**Comment by riccardobrasca** [comment]
> > Is the other implication already in mathlib?
> 
> Yes. It is [`IsDedekindDomain.isDedekindDomainDvr`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/DedekindDomain/Dvr.html#IsDedekindDomain.isDedekindDomainDvr).

OK, so the TODO in [IsDedekindDomainDvr](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/DedekindDomain/Dvr.html#IsDedekindDomainDvr) can be removed.

Also, I think that in Lean4 both structures can be classes (with instances in both directions). Can you try and see if this works (maybe in another PR)? Thanks!

bors d+

**Comment by riccardobrasca** [comment]
Let's just wait for the benching results (I don't think anything strange will happen, but who knows).

!bench

**Comment by riccardobrasca** [comment]
!bench

**Pull request closed without merging**